File: src/main/java/me/itzisonn_/meazy/parser/operator/Operator.java
package me.itzisonn_.meazy.parser.operator;

import lombok.Getter;
import me.itzisonn_.meazy.runtime.value.RuntimeValue;

/**
 * Gives ability to create expressions faster
 */
@Getter
public abstract class Operator {
    /**
     * Symbol
     */
    private final String symbol;
    /**
     * Operator type
     */
    private final OperatorType operatorType;
    /**
     * Whether this operator is overridable by operator functions
     */
    private final boolean isOverridable;

    /**
     * Main constructor
     *
     * @param symbol Symbol
     * @param operatorType Operator type
     * @param isOverridable Whether this operator is overridable by operator functions
     *
     * @throws NullPointerException If either symbol or operatorType is null
     */
    public Operator(String symbol, OperatorType operatorType, boolean isOverridable) throws NullPointerException {
        if (symbol == null) throw new NullPointerException("Symbol can't be null");
        if (operatorType == null) throw new NullPointerException("OperatorType can't be null");

        this.symbol = symbol;
        this.operatorType = operatorType;
        this.isOverridable = isOverridable;
    }

    /**
     * Constructor with isOverridable set to true
     *
     * @param symbol Symbol
     * @param operatorType Operator type
     *
     * @throws NullPointerException If either symbol or operatorType is null
     */
    public Operator(String symbol, OperatorType operatorType) throws NullPointerException {
        this(symbol, operatorType, true);
    }

    /**
     * Calculates expression value with this operator
     *
     * @param value1 First value
     * @param value2 Second value or null if this operator type isn't {@link OperatorType#INFIX}
     * @return Resulted value
     */
    public abstract RuntimeValue<?> calculate(RuntimeValue<?> value1, RuntimeValue<?> value2);
}

File: src/main/java/me/itzisonn_/meazy/parser/operator/Operator.kt
package me.itzisonn_.meazy.parser.operator;

import me.itzisonn_.meazy.instruction.InstructionsSet;
import me.itzisonn_.meazy.parser.DataType;
import me.itzisonn_.meazy.parser.ast.expression.OperatorExpression;
import me.itzisonn_.meazy.runtime.environment.Environment;
import org.jspecify.annotations.NullMarked;

/**
 * Gives ability to create expressions faster
 */
@NullMarked
public abstract class Operator {
    /**
     * Id
     */
    private final String id;
    /**
     * Symbol
     */
    private final String symbol;
    /**
     * Operator type
     */
    private final OperatorType operatorType;
    /**
     * Whether this operator is overridable by operator functions
     */
    private final boolean isOverridable;

    /**
     * Main constructor
     *
     * @param id            Id
     * @param symbol        Symbol
     * @param operatorType  Operator type
     * @param isOverridable Whether this operator is overridable by operator functions
     */
    public Operator(String id, String symbol, OperatorType operatorType, boolean isOverridable) {
        this.id = id;
        this.symbol = symbol;
        this.operatorType = operatorType;
        this.isOverridable = isOverridable;
    }

    /**
     * Constructor with isOverridable set to true
     *
     * @param id           Id
     * @param symbol       Symbol
     * @param operatorType Operator type
     */
    public Operator(String id, String symbol, OperatorType operatorType) {
        this(id, symbol, operatorType, true);
    }

    /**
     * Calculates expression value with this operator
     *
     * @param instructionsSet    InstructionsSet
     * @param environment        Environment
     * @param operatorExpression Operator expression
     */
    public abstract void emit(InstructionsSet instructionsSet, Environment environment, OperatorExpression operatorExpression);

    /**
     * TODO
     *
     * @param environment        Environment
     * @param operatorExpression Operator expression
     */
    public abstract DataType getType(Environment environment, OperatorExpression operatorExpression);

    public String getId() {
        return this.id;
    }

    public String getSymbol() {
        return this.symbol;
    }

    public OperatorType getOperatorType() {
        return this.operatorType;
    }

    public boolean isOverridable() {
        return this.isOverridable;
    }
}
